sklearn/partial_dependence.py

```diff
@@ -0,0 +1,577 @@
+"""Partial dependence plots for regression and classification models."""
+
+# Authors: Peter Prettenhofer
+#          Trevor Stephens
+#          Nicolas Hug
+# License: BSD 3 clause
+
+from itertools import count
+import numbers
+import warnings
+
+import numpy as np
+from scipy.stats.mstats import mquantiles
+
+from .base import is_classifier, is_regressor
+from .utils.extmath import cartesian
+from .externals.joblib import Parallel, delayed
+from .externals import six
+from .externals.six.moves import map, range, zip
+from .utils import check_array
+from .utils.validation import check_is_fitted
+from .tree._tree import DTYPE
+from .exceptions import NotFittedError
+from .ensemble.gradient_boosting import BaseGradientBoosting
+from .ensemble._gradient_boosting import _partial_dependence_tree
+
+
+__all__ = ['partial_dependence', 'plot_partial_dependence']
+
+
+def _grid_from_X(X, percentiles=(0.05, 0.95), grid_resolution=100):
+    """Generate a grid of points based on the ``percentiles of ``X``.
+
+    The grid is a cartesian product between the columns of Z. The ith column of
+    Z consists in ``grid_resolution`` equally-spaced points between the
+    percentiles of the ith column of X.
+    If ``grid_resolution`` is bigger than the number of unique values in the
+    ith column of X, then those unique values will be used instead.
+
+    Parameters
+    ----------
+    X : ndarray
+        The data
+    percentiles : tuple of floats
+        The percentiles which are used to construct the extreme values of
+        the grid.
+    grid_resolution : int
+        The number of equally spaced points to be placed on the grid for a
+        given column.
+
+    Returns
+    -------
+    grid : ndarray, shape=(n_points, X.shape[1])
+        All data points on the grid. n_points is always ``<= grid_resolution **
```

Comment:
in what case is n_points strictly less?

Reply:
I think this should be: A value for each feature at each point in the grid.

Reply:
`n_points` is strictly less when one of the features has less than `grid_resolution` unique values. The correct formula for `n_points` is

```python
n_points = prod(min(grid_resolution, n_unique_values(X[:, i]))
                for i in range(X.shape[1]))`
```

but I find it a bit overkill.